### PR TITLE
Added style AST to parser services

### PR DESCRIPTION
**Pull request by marekdedic**
Alternative proposal to #318, see https://github.com/sveltejs/svelte-eslint-parser/pull/318#issuecomment-1521005109

### CODE REVIEW

File: src/parser/index.ts
Comment (marekdedic) [inline src/parser/index.ts:259]: @ota-meshi What do you think is the right thing to do here? Probably not using the console?

File: src/parser/index.ts
Comment (marekdedic) [inline src/parser/index.ts:177]: I am thinking that maybe it would be useful to have a distinction between 2 states, that currently both produce `null` as a result:
1. There is no `SvelteStyleElement`
2. There is one, but the AST couldn't be parsed

@ota-meshi what do you think? I am not sure what is the right approach here...

Comment (ota-meshi) [reply src/parser/index.ts:177]: I think it makes more sense to return more information together. I also think it useful to return information whether it was parsed as CSS or SCSS.

File: src/parser/index.ts
Comment (ota-meshi) [inline src/parser/index.ts:270]: In case of CSS parse errors, I find it useful to return that information. You can add a new rule to display parse errors as warnings. Rules can also optionally use `postcss-safe-parser` to fix parsing errors.
https://www.npmjs.com/package/postcss-safe-parser

Comment (ota-meshi) [reply src/parser/index.ts:270]: (However, I think using stylelint makes more sense than creating a new CSS parse error rule.)

Comment (ota-meshi) [inline src/parser/index.ts:270]: Does the node's location return the location of the svelte file?
I think it makes more sense to return the correct location in advance rather than having the rule calculate the offset location.

File: src/parser/index.ts
Comment (marekdedic) [reply src/parser/index.ts:177]: That's a great idea, thanks!

File: src/parser/index.ts
Comment (marekdedic) [reply src/parser/index.ts]: You mean errors produced by SCSS? Hmm, I am not sure what would happen currently, maybe it would all crash? I'll investigate more...

Comment (ota-meshi) [reply src/parser/index.ts]: Changing the offset later would be a breaking change, so could you include it in this PR?

Comment (ota-meshi) [reply src/parser/index.ts]: I think that it is good to do what is calculated by `convertPostCSSNodeToESLintNode`. But, we can assign directly to `node.source`.

Comment (ota-meshi) [reply src/parser/index.ts]: I'm talking about CSS syntax errors.
I don't think we should stop building the entire AST just because svelte file has CSS with syntax errors. I think we need to handle syntax errors with `try {} catch {}`.
I think it would be nice to add to the information in getStyleContext if there were any syntax errors.

Comment (marekdedic) [reply src/parser/index.ts]: I'd like to add the parse errors, however, I am currently blocked by postcss/postcss#1839 :(

Comment (marekdedic) [reply src/parser/index.ts]: Well, see the comments in the linked postcss issue - IMHO it's not worth the added complexity to list parse errors.

File: src/parser/style-context.ts
Comment (ota-meshi) [inline src/parser/style-context.ts:11]: I think it's better to change the type to the following to make the state easier to understand. What do you think?

```suggestion
export interface StyleContextSuccess {
  sourceLang: string | null;
  sourceAst: Root;
  error: false;
  unknown?: false;
}
export interface StyleContextError {
  sourceLang: string | null;
  sourceAst: null;
  error: true;
  unknown?: false;
}
export interface StyleContextNone {
  sourceLang: null;
  sourceAst: null;
  error?: false;
  unknown?: false;
}
export interface StyleContextUnknown {
  sourceLang: string;
  sourceAst: null;
  error?: false;
  unknown: true;
}
export type StyleContext = StyleContextSuccess | StyleContextError | StyleContextNone | StyleContextUnknown
```

I think the new rule should stop all reports if `StyleContextError`, or `StyleContextUnknown`.

File: src/parser/style-context.ts
Comment (ota-meshi) [inline src/parser/style-context.ts:85]: For `<style>` tags like below, I think we also need to change the `column` position.

```svelte
<style> a { color: red; } </style>
```

**Review by ota-meshi — CHANGES_REQUESTED** [review]

File: src/parser/index.ts
Comment (ota-meshi) [reply src/parser/index.ts]: I think that it is enough if we can get one error information :wink:

File: src/parser/style-context.ts
Comment (marekdedic) [reply src/parser/style-context.ts:11]: Great idea!
I modified it a bit to be a proper discriminated union (mostly so `if (context.status === "success")` would be a happy-path for typing), please take a look and let me know if I missed anything ;)

File: src/parser/style-context.ts
Comment (marekdedic) [reply src/parser/style-context.ts:104]: Oh no...

File: src/parser/style-context.ts
Comment (ota-meshi) [reply src/parser/style-context.ts]: It looks good to me! 👍

File: src/parser/style-context.ts
Comment (marekdedic) [reply src/parser/style-context.ts:101]: However I did fix the issue you reported. I just don't know if it even makes sense to try to offset things... :/

Comment (ota-meshi) [reply src/parser/style-context.ts:90]: Thanks for your opinion.
As you said, I think the difference in range specifications can be confusing.
However, I don't want to write calculation logic each time a new rule in the future requests to create a report from a PostCSS node.
What do you think about providing a helper function to get the report range from PostCSS nodes via the parser service instead of reassigning all node positions?

Comment (marekdedic) [reply src/parser/style-context.ts:130]: That's a good idea, I have implemented a pair of functions that does that and tested them.

However, I don't know how to export them - should they be a normal export from the package? So you would (as a user) do just 
```ts
import { styleNodeLoc, styleNodeRange } from "svelte-eslint-parser";

styleNodeLoc(services.getStyleContext().sourceAst!);
```
Or should they be properties on `StyleContext`? So you would do
```ts
const styleContext = services.getStyleContext();
styleContext.styleNodeLoc(styleContext.sourceAst!)
```

Which do you think is better? I noticed that you don't export many things directly...

Also, should the `StyleContext` type (or the subtypes) be exported?

Comment (ota-meshi) [reply src/parser/style-context.ts:130]: Sorry for the late reply. I've been busy with office work lately 😓 
I was thinking of providing it from a parser service. So the second example is the same.

Comment (ota-meshi) [reply src/parser/style-context.ts:130]: > Also, should the `StyleContext` type (or the subtypes) be exported?

That's a good point. I think it makes sense to expose `StyleContext` type so that we can write rules in TypeScript.

Comment (marekdedic) [reply src/parser/style-context.ts:130]: No worries about the time :)

I have added `styleNodeLoc` and `styleNodeRange` to the parser services and exported the types around `StyleContext`.

**Review by ota-meshi — APPROVED** [review]
LGTM! I think we can merge once the CI error is fixed. I think maybe there is a problem with the main branch.

**Comment by marekdedic** [comment]
Thank you for the reviews, re-reviews and consultation of the correct approach! :)

**Pull request merged**
